File: common/src/main/java/com/pluxity/category/entity/Category.java
package com.pluxity.category.entity;

import com.pluxity.global.constant.ErrorCode;
import com.pluxity.global.entity.BaseEntity;
import com.pluxity.global.exception.CustomException;
import jakarta.persistence.*;
import java.util.ArrayList;
import java.util.List;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "category")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Category extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private CategoryType type;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_id")
    private Category parent;

    @OneToMany(mappedBy = "parent")
    private List<Category> children = new ArrayList<>();

    @Builder
    public Category(String name, CategoryType type, Category parent) {
        this.name = name;
        this.type = type;
        assignToParent(parent);
        validateDepth();
    }

    public boolean isRoot() {
        return parent == null;
    }

    public int getDepth() {
        return isRoot() ? 1 : parent.getDepth() + 1;
    }

    public void assignToParent(Category newParent) {
        validateParentType(newParent);
        
        if (this.parent != null) {
            this.parent.children.remove(this);
        }

        this.parent = newParent;

        if (newParent != null) {
            newParent.children.add(this);
        }

        validateDepth();
    }

    public void updateName(String name) {
        this.name = name;
    }

    private void validateParentType(Category newParent) {
        if (newParent != null && !this.type.equals(newParent.type)) {
            throw new CustomException(ErrorCode.INVALID_CATEGORY_TYPE);
        }
    }

    private void validateDepth() {
        if (getDepth() > type.getMaxDepth()) {
            throw new CustomException(ErrorCode.EXCEED_CATEGORY_DEPTH);
        }
    }
}

File: common/src/main/java/com/pluxity/category/entity/Category.kt
package com.pluxity.category.entity

import com.pluxity.global.constant.ErrorCode
import com.pluxity.global.entity.BaseEntity
import com.pluxity.global.exception.CustomException
import jakarta.persistence.Column
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.ManyToOne
import jakarta.persistence.MappedSuperclass
import jakarta.persistence.OneToMany

@MappedSuperclass
abstract class Category<T : Category<T>> : BaseEntity() {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    open var id: Long? = null

    @Column(nullable = false)
    open var name: String = ""

    @ManyToOne(fetch = FetchType.LAZY)
    open var parent: T? = null

    @OneToMany(mappedBy = "parent", orphanRemoval = true)
    open var children: MutableList<T> = mutableListOf()

    open val maxDepth: Int
        get() = 2

    val depth: Int
        get() = 1 + (parent?.depth ?: 0)

    fun isRoot(): Boolean = parent == null

    fun updateName(name: String) {
        this.name = name
    }

    @Suppress("UNCHECKED_CAST")
    fun assignToParent(newParent: T?) {
        val currentParent = this.parent
        if (currentParent != null) {
            currentParent.children.remove(this as T)
        }

        this.parent = newParent

        if (newParent != null) {
            newParent.children.add(this as T)
        }

        validateDepth()
    }

    fun validateDepth() {
        if (depth > maxDepth) {
            throw CustomException(ErrorCode.EXCEED_CATEGORY_DEPTH)
        }
    }
}
